source/compiler/qsc_qasm/src/semantic/lowerer.rs

```diff
@@ -4118,6 +4130,48 @@ impl Lowerer {
         self.lower_indexed_ident_expr_rec(collection, indices.into())
     }
 
+    fn change_index_endianness(&mut self, idx: semantic::Index, width: u32) -> semantic::Index {
+        // If we have a variable `var` of a 32-bit type, and we want to change
+        // the endianness of the index when doing var[idx], we need to do var[32 - idx - 1].
+        let width_minus_expr_minus_one = |expr: Expr| -> Expr {
+            let Some(semantic::LiteralKind::Int(mut val)) = expr.const_value else {
+                unreachable!("range components are guaranteed to be const exprs");
+            };
+            // OpenQASM indexing is n-based, so we wrap around negative indices.
+            while val < 0 {
+                val += i64::from(width);
+            }
+            Expr::int(i64::from(width) - val - 1, expr.span)
+        };
+
+        match idx {
+            semantic::Index::Expr(expr) => {
+                let new_expr = if expr.ty.is_const() {
+                    width_minus_expr_minus_one(expr.with_const_value(self))
+                } else {
+                    Expr::bin_op(
+                        semantic::BinOp::Sub,
+                        // Instead of building two nested bin_ops, we just substract 1 from width.
+                        Expr::int(i64::from(width - 1), expr.span),
+                        expr,
+                    )
+                };
+                semantic::Index::Expr(new_expr)
+            }
+            semantic::Index::Range(range) => {
+                // When changing the endianness of a range, we need to swap the start and end.
+                let start = range.end.map(width_minus_expr_minus_one);
+                let end = range.start.map(width_minus_expr_minus_one);
+                semantic::Index::Range(Box::new(semantic::Range {
+                    span: range.span,
+                    start,
+                    end,
+                    step: range.step,
+                }))
```

Comment:
I'm still a little confused on how `step` factors in here. Does it also need to be transformed, or can we safely assume that the step is still correct even after this transformation has taken place?